stack/stack_funcs.cpp

```diff
@@ -0,0 +1,78 @@
+#include "stack_funcs.h"
+
+void stack_ctor(my_stack *stk) {
+    stk->canary_start = CANARY_VALUE;
+    stk->canary_end = CANARY_VALUE;
+    stk->data = (stackelem_t *)calloc(2, sizeof(stackelem_t));
+    stk->data[0] = CANARY_VALUE;
+    stk->size = 1; // 1 = 0
+    stk->capacity = reall_factor;
+    stk->data[stk->capacity - 1] = CANARY_VALUE;
+    stk->popped = NAN;
+    stk->hash = hash_struct(stk);
+    printf("ctor!\n");
+    stack_dump(stk); 
+}
+
+my_stack *realloc_incr_stack(my_stack *stk) {
+    stk->capacity = stk->capacity*reall_factor;
+    stk->data = (stackelem_t *)realloc(stk->data, stk->capacity*sizeof(stackelem_t));
+    stk->data[stk->capacity - 1] = CANARY_VALUE;
+    for (int j = stk->size + 1; j < stk->capacity - 1; j++) {
+        stk->data[j] = NAN; //init NAN like posion and delete old canary
+    }
+    return stk;
+}
+
+my_stack *realloc_decr_stack(my_stack *stk) {
+    stk->capacity = stk->capacity/reall_factor;
+    stk->data = (stackelem_t *)realloc(stk->data, stk->capacity*sizeof(stackelem_t));
+    stk->data[stk->capacity - 1] = CANARY_VALUE;
+    return stk;
+}
+
+void stack_push(my_stack *stk, stackelem_t elem) { //newest in end
```

Comment:
why return value is `void`? how to check whether push was successful or not?

Reply:
because stack_error in functions is the flag of correct work